HTTPCORE-759 Content-Length missing on null entity request content

chrysophylax opened the pull request:
Quick fix for maximum compatibility per JIRA

ok2c commented on httpcore5/src/main/java/org/apache/hc/core5/http/protocol/RequestContent.java:
```diff
@@ -110,6 +110,10 @@ public void process(final HttpRequest request, final EntityDetails entity, final
                 throw new ProtocolException("Content-Length header already present");
             }
         }
+        if (entity == null) {
```
@chrysophylax I think this should be done for `POST`, `PUT` and `PATCH` only and not any other standard method. That would be consistent with the behavior of HC 4.x as well.

reschke replied:
I'd actually use a denyList with those methods that are documented not to anticipate content (which are only few).

chrysophylax replied:
Isn't a HttpEntityEnclosingRequest inherently only for these methods which the standard defines Enclosed Content semantics?

ok2c replied:
@chrysophylax We had to drop `HttpEntityEnclosingRequest` in 5.x because RFC 723x muddied the water by declaring it was legal for all methods to enclose an entity, well, except for `TRACE`, that is.

reschke replied:
RFC 723x did not "muddy the water" - it just documents how HTTP works in practice.

chrysophylax replied:
Ahhh. I follow. 

@reschke are there more methods that will not anticipate content bar `TRACE`? I could modify it either way (only `POST`, `PUT`, `PATCH`) or a denyList (`TRACE`, `??`)

ok2c replied:
@chrysophylax Please do `POST`, `PUT`, `PATCH` by default. If desired the users can always override the default behavior through a custom request interceptor.

reschke replied:
I believe TRACE is the only method that "MUST NOT" have content. There are some ("GET" and "DELETE"?) which "SHOULD NOT", so for these I wouldn't add "Content-Length: 0", either.

ok2c commented:
@chrysophylax Please also make sure the build still passes all style checks. I would also kindly ask you to add a test case to `TestStandardInterceptors` for the new behavior.

chrysophylax commented:
I updated the existing `testRequestContentNullEntity` test  to the new expected behaviour and added a test to make sure non-PUT/POST/PATCH don't get the `Content-Length` header added.

reschke reviewed (commented):
I still don't get why we're making those special; I would reverse the logic.
